### PR TITLE
fix: add 'CONTAIN' support in query builder

### PR DESCRIPTION
**Pull request by yangyangv2**
This fix to address the issue https://github.com/linkedin/datahub-gma/issues/114

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java [hunk 1]
```diff
@@ -78,6 +78,8 @@ public static QueryBuilder getQueryBuilderFromCriterion(@Nonnull Criterion crite
       return QueryBuilders.rangeQuery(criterion.getField()).lt(criterion.getValue().trim());
     } else if (condition == Condition.LESS_THAN_OR_EQUAL_TO) {
       return QueryBuilders.rangeQuery(criterion.getField()).lte(criterion.getValue().trim());
+    } else if (condition == Condition.CONTAIN) {
+      return QueryBuilders.termQuery(criterion.getField(), criterion.getValue().trim());
```
Comment (keremsahin1) [inline dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:82]: I don't think term query is fitting CONTAIN use case, match query is a better fit.

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html
https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html

Comment (jywadhwani) [reply dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:82]: My thoughts on this:
First let us define what the desired behavior is. Does `CONTAIN` search for a substring of a word, or only a prefix? If the field that the `Criterion` model contains has been analyzed such that it indexes those substring tokens then match query should suffice, but what if the field doesn't?
Here is an example of using prefix based queries in the search request vs. indexing those tokens (https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-prefix-query.html)
The former approach is slow, but is also agnostic to the mapping that exists.

Comment (keremsahin1) [reply dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:82]: I think prefix is another condition `START_WITH`.

Comment (yangyangv2) [reply dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:82]: @keremsahin1  it's a good call out. It's should be match query

@jywadhwani, there're also `START_WITH` and `END_WITH` defined in the `Criterion`, which we can fit `prefix query` or `suffix query` respectively. I'm thinking we just treat `CONTAIN` as simple text query as general understanding, and use 'prefix' and 'suffix query' as other Criterion, what do you think? Besides, I'm also confirming this behavior with the issue reporter, still waiting for response.

Comment (jywadhwani) [reply dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:82]: Looks like the reporter expects the desired behavior as "Contains operator should work for substring of string fields. " as mentioned in the issue itself.

Comment (yangyangv2) [reply dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:82]: Just updated the PR, can you take a look again? @jywadhwani @keremsahin1 . Thanks!

**Comment by keremsahin1** [comment]
There seems to be something off here. Why do I see diff from @kaliang1 in this PR?

File: dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java (hunk 1 above)
Comment (keremsahin1) [reply dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:82]: I though you changed this to `match` query @yangyangv2 ?

Comment (yangyangv2) [reply dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java]: you're right. there was a duplicated implementation... Just updated the PR.

File: dao-impl/elasticsearch-dao/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java [hunk 2]
```diff
@@ -88,8 +88,9 @@ public static QueryBuilder getQueryBuilderFromCriterion(@Nonnull Criterion crite
       return QueryBuilders.rangeQuery(criterion.getField()).lt(criterion.getValue().trim());
     } else if (condition == Condition.LESS_THAN_OR_EQUAL_TO) {
       return QueryBuilders.rangeQuery(criterion.getField()).lte(criterion.getValue().trim());
+    } else if (condition == Condition.CONTAIN) {
+      return QueryBuilders.matchQuery(criterion.getField(), criterion.getValue().trim());
```
Comment (jywadhwani) [inline dao-impl/elasticsearch-dao/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:92]: should this be a `wildcard` query instead? https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-wildcard-query.html

File: dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java [hunk 3]
```diff
@@ -60,6 +60,10 @@ private SearchUtils() {
    * <p>If the condition between a field and value is not the same as EQUAL, a Range query is constructed. This
    * condition does not support multiple values for the same field.
    *
+   * <p>When CONTAIN, START_WITH and END_WITH conditions are used, the underlying logic is using wildcard query which is
+   * not performant according to ES. For details, please refer to:
+   * https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-wildcard-query.html#wildcard-query-field-params
```
Comment (camelliazhang) [inline dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:65]: LGTM in general; Regarding performance, how do we ensure this is not stressing the server (in some extreme cases)?

**Review by camelliazhang — COMMENTED** [review]
also, the current query template supports advanced query as well, so that user can just issue *text? https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html

File: dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java (hunk 3 above)
Comment (yangyangv2) [reply dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java:65]: it's a known issue that wildcast query is running known. For extreme case consideration, we could add a timeout for wildcard query and include it into the SLA. 

At the moment, this is only requested from OSS, we can just push the update and keep an eye on the feedback how it performs in real-life cases.

File: dao-impl/elasticsearch-dao/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java (hunk 2 above)
Comment (yangyangv2) [reply dao-impl/elasticsearch-dao/src/main/java/com/linkedin/metadata/dao/utils/SearchUtils.java]: yes, according to this [decision tree](https://www.elastic.co/blog/find-strings-within-strings-faster-with-the-new-elasticsearch-wildcard-field), wildcard is the right choice here.

**Comment by yangyangv2** [comment]
> also, the current query template supports advanced query as well, so that user can just issue *text? https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html

Good to know! It could be good for advance users.

This specific ask is for supporting CONTAIN with sbustring query, we prefer to have explicit filter options (CONTAIN, START_WITH, END_WITH) to users and hide ES internal details such as '*'. If users knows they are looking for START_WITH, then using START_WITH will perform better as the underlying query pattern is _wildcard_query(term*)_. 

Exposing '*' to users may lead to complexity and out of controlled use cases such as searching on "*A*B*C*D*"..etc.

**Comment by yangyangv2** [comment]
> also, the current query template supports advanced query as well, so that user can just issue *text? https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html

**Review by keremsahin1 — APPROVED** [review]

**Pull request merged**